Add user name fetching and assignment functionality (#242)

tansdf opened the pull request:
- Introduced a centralized async method `fetch_and_assign_user_name` in the Application class to fetch user details and assign the full name to incidents.
- Updated Mattermost, Slack, and Telegram application classes to utilize the new method for user name assignment, improving efficiency by leveraging cached user data.
- Added a caching mechanism in the Incidents class to avoid redundant API calls for user name lookups.

This enhancement streamlines user name handling across different messaging platforms, ensuring a more responsive user experience.

tansdf committed:
Add user name fetching and assignment functionality (#242)

- Introduced a centralized async method `fetch_and_assign_user_name` in the Application class to fetch user details and assign the full name to incidents.
- Updated Mattermost, Slack, and Telegram application classes to utilize the new method for user name assignment, improving efficiency by leveraging cached user data.
- Added a caching mechanism in the Incidents class to avoid redundant API calls for user name lookups.

This enhancement streamlines user name handling across different messaging platforms, ensuring a more responsive user experience.

tansdf committed:
Remove unused user name variable from MattermostApplication class

DiTsi committed:
Add incident notifications settings to configuration documentation (#270)

- Introduced new section for incident.notifications in config_file.md
- Added incident.notifications.assignment

DiTsi commented on app/im/mattermost/mattermost_application.py:
```diff
@@ -56,11 +56,15 @@ async def get_user_details(self, user_details):
-                return {'id': id_, 'username': data.get('username'), 'exists': exists}
+                    first_name = data.get('first_name', '').strip()
+                    last_name = data.get('last_name', '').strip()
+                    full_name = f"{first_name} {last_name}".strip() or data.get('username')
```
`full_name` always contains first name and last name.
We need an additional variable `username` that contains the username.

`full_name` is used only in the UI, while `username` is used in messenger integration.

DiTsi commented on app/im/slack/slack_application.py:
```diff
@@ -48,11 +48,15 @@ async def get_user_details(self, user_details):
-                return {'id': id_, 'exists': exists}
+                    user_data = data.get('user', {})
+                    profile = user_data.get('profile', {})
+                    full_name = profile.get('real_name') or profile.get('display_name') or user_data.get('name')
```
we should use only `real_name_normalized`

DiTsi commented on app/im/telegram/telegram_application.py:
```diff
@@ -247,19 +243,24 @@ def _markdown_links_to_native_format(self, text):
+                        chat_data = data.get('result', {})
+                        first_name = chat_data.get('first_name', '').strip()
+                        last_name = chat_data.get('last_name', '').strip()
+                        full_name = f"{first_name} {last_name}".strip() or chat_data.get('username')
```
we should use `first_name` + `last_name` only

DiTsi reviewed (changes requested):
`incident_.assign_user(user_name)` should work as before.
But we need to add `incident_.assign_fullname(user_fullname)`.
The incident should have a new property `assigned_fullname`.

tansdf committed:
Refactor user details fetching in Mattermost, Slack, and Telegram applications

- Simplified the `get_user_details` method in MattermostApplication, SlackApplication, and TelegramApplication classes by removing unnecessary checks for user ID.
- Improved readability and efficiency of the user details retrieval process across all messaging platforms.
- Added a new method `get_assigned_user_by_id` to the Incidents class to cache user details.
- Added a new field `assigned_user_fullname` to the Incident class to store the full name of the assigned user.
- Added a new field `assigned_user` to the Incident class to store the username of the assigned user.
- Added a new method `assign_fullname` to the Incident class to assign the full name of the user.
- Added a new method `assign_user` to the Incident class to assign the username of the user.

tansdf committed:
Fix variable name for assigned user fullname in Incident class

tansdf committed:
Implement assignment notification feature across messaging platforms (#269)

* Implement assignment notification feature across messaging platforms

- Added a new configuration option for incident notifications in `config.py`.
- Introduced `post_assignment_notification` method in the Application class to handle user assignment notifications.
- Implemented `format_user_mention` method in Mattermost, Slack, and Telegram application classes to format user mentions appropriately.
- Updated incident handling to trigger notifications upon user assignment, enhancing user engagement and responsiveness.

This update improves the incident management experience by notifying users when they are assigned to incidents.

* Add user name retrieval in MattermostApplication class

- Introduced a new variable `user_name` to fetch the user's name from the payload.
- This addition supports enhanced user identification in incident handling processes.

This update lays the groundwork for improved user engagement by incorporating user name data into the application logic.

* Refactor post_assignment_notification method and user mention formatting

- Simplified the `post_assignment_notification` method by removing the `incidents` and `user_display_name` parameters, streamlining the notification process.
- Updated the `format_user_mention` method in Mattermost and Telegram applications to directly use the `display_name` parameter, enhancing user mention formatting.
- Adjusted the invocation of `post_assignment_notification` across Mattermost, Slack, and Telegram applications to reflect the updated method signature.

These changes improve code clarity and maintainability while ensuring consistent user notifications across messaging platforms.

* Fix telegram missing display_name

tansdf committed:
Update TelegramApplication to include username in user details response

tansdf committed:
Assign user display name to incident upon user retrieval in TelegramApplication
